x/ccv/provider/keeper/distribution.go

```diff
@@ -57,32 +68,225 @@ func (k Keeper) GetAllConsumerRewardDenoms(ctx sdk.Context) (consumerRewardDenom
+	store := ctx.KVStore(k.storeKey)
+	b := store.Get(types.ConsumerRewardsAllocationKey(chainID))
+	k.cdc.MustUnmarshal(b, &pool)
+	return
+}
+
+// SetConsumerRewardsAllocation sets the consumer rewards allocation for the given chain ID
+func (k Keeper) SetConsumerRewardsAllocation(ctx sdk.Context, chainID string, pool types.ConsumerRewardsAllocation) {
+	store := ctx.KVStore(k.storeKey)
+	b := k.cdc.MustMarshal(&pool)
+	store.Set(types.ConsumerRewardsAllocationKey(chainID), b)
+}
+
+// GetConsumerRewardsPool returns the balance
+// of the consumer rewards pool module account
+func (k Keeper) GetConsumerRewardsPool(ctx sdk.Context) sdk.Coins {
+	return k.bankKeeper.GetAllBalances(
+		ctx,
+		k.accountKeeper.GetModuleAccount(ctx, types.ConsumerRewardsPool).GetAddress(),
+	)
+}
+
+// ComputeConsumerTotalVotingPower returns the validator set total voting power
+// for the given consumer chain
+func (k Keeper) ComputeConsumerTotalVotingPower(ctx sdk.Context, chainID string) (totalPower int64) {
+	// sum the opted-in validators set voting powers
+	for _, v := range k.GetConsumerValSet(ctx, chainID) {
+		totalPower += v.Power
+	}
+
+	return
+}
+
+// IdentifyConsumerChainIDFromIBCPacket checks if the packet destination matches a registered consumer chain.
+// If so, it returns the consumer chain ID, otherwise an error.
+func (k Keeper) IdentifyConsumerChainIDFromIBCPacket(ctx sdk.Context, packet channeltypes.Packet) (string, error) {
+	channel, ok := k.channelKeeper.GetChannel(ctx, packet.DestinationPort, packet.DestinationChannel)
+	if !ok {
+		return "", errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "channel not found for channel ID: %s", packet.DestinationChannel)
+	}
+	if len(channel.ConnectionHops) != 1 {
+		return "", errorsmod.Wrap(channeltypes.ErrTooManyConnectionHops, "must have direct connection to consumer chain")
 	}
+	connectionID := channel.ConnectionHops[0]
+	_, tmClient, err := k.getUnderlyingClient(ctx, connectionID)
+	if err != nil {
+		return "", err
+	}
+
+	chainID := tmClient.ChainId
+	if _, ok := k.GetChainToChannel(ctx, chainID); !ok {
+		return "", errorsmod.Wrapf(types.ErrUnknownConsumerChannelId, "no CCV channel found for chain with ID: %s", chainID)
+	}
+
+	return chainID, nil
+}
+
+// HandleSetConsumerCommissionRate sets a per-consumer chain commission rate for the given provider address
+// on the condition that the given consumer chain exists.
+func (k Keeper) HandleSetConsumerCommissionRate(ctx sdk.Context, chainID string, providerAddr types.ProviderConsAddress, commissionRate sdk.Dec) error {
+	// check that the consumer chain exists
+	if !k.IsConsumerProposedOrRegistered(ctx, chainID) {
+		return errorsmod.Wrapf(
+			types.ErrUnknownConsumerChainId,
+			"unknown consumer chain, with id: %s", chainID)
+	}
+	// set per-consumer chain commission rate for the validator address
+	k.SetConsumerCommissionRate(
```

Comment:
Do we want to have a minimum consumer commission rate, like we have on the Hub?

Reply:
Seems like a good idea; should probably just use the same value rather than doing smth more complicated like having a different minimal rate per consumer?

Reply:
I will try to do this one

Reply:
Done, added also some extra validation on the rate which we didn't have before:
https://github.com/cosmos/interchain-security/pull/1834/files